### DISCUSSION TITLE
Code that allocates dynamic memory with malloc function and then reallocates 4 bytes using realloc function

### DISCUSSION THREAD

**Discussion by wldhks1959** [community/community#38622] [Other Feature Feedback, Questions, & Ideas]
'#include<stdio.h>
#include<stdlib.h>
int main(void){
 int i=0;
 int *p=(int*)malloc(sizeof(int)*2);
 p[0]=10;
 p[1]=20;

 p=(int*)realloc(p,sizeof(int)*1);
 p[0]=30;

 for(i=0;i<2;i++)
  printf("p[%i] : %d \n", i, p[i]);

free(p);
p=NULL;
return 0; 
}`

The result of the execution
p[0] : 30 
p[1] : 20 

I wrote the code like this..
After allocating 8 bytes, I reallocated it to 4 bytes, so the value of p[1] should be garbage, so why did we get 20?

**Comment by airtower-luna** [comment]
What you're doing there is called "out of bounds read": You're reading memory beyond the (new) end of an array, memory your application doesn't own any more. Behavior in that case is undefined (and often a security problem!), as it happens you're lucky and get the value previously written to the area instead of a crash.

**Reply by wldhks1959** [reply]
Thank you for your reply. I understand. In conclusion, it was reallocated, but it was printed even after the previously allocated p[1] value was reallocated.

**Reply by airtower-luna** [reply]
The *value* wasn't changed. With `realloc() your program gave up ownership of the memory. The memory itself is still there, and *happened* to still contain the same value. Mind that you can't rely on that behavior, and should be very careful to make sure your software *never* does anything like this. Most security issues are caused by invalid memory access (usually by accident).
